pkg/agent/openflow/pipeline.go

```diff
@@ -1145,123 +1139,169 @@ func (c *client) localProbeFlow(localGatewayIP net.IP, category cookie.Category)
-			SNAT(snatIPRange, nil).
-			LoadToMark(snatCTMark).CTDone().
-			Cookie(c.cookieAllocator.Request(category).Raw()).
-			Done(),
 	}
 	return flows
 }
 
-func (c *client) l3ToExternalFlows(nodeIP net.IP, localSubnet net.IPNet, outputPort int, category cookie.Category) []binding.Flow {
+// snatFlows installs flows to perform SNAT for traffic to the external network.
+// It is used on Windows Nodes only.
+func (c *client) snatFlows(nodeIP net.IP, localSubnet net.IPNet, category cookie.Category) []binding.Flow {
+	snatIPRange := &binding.IPRange{StartIP: nodeIP, EndIP: nodeIP}
+	l3FwdTable := c.pipeline[l3ForwardingTable]
+	nextTable := l3FwdTable.GetNext()
 	flows := []binding.Flow{
-		// Forward the packet to L2ForwardingCalc table if it is communicating to a Service.
-		c.pipeline[l3ForwardingTable].BuildFlow(priorityNormal).
+		// First install flows for traffic that should bypass SNAT.
+
+		// This flow is for traffic to the local Pod subnet.
+		l3FwdTable.BuildFlow(priorityNormal).
 			MatchProtocol(binding.ProtocolIP).
 			MatchRegRange(int(marksReg), markTrafficFromLocal, binding.Range{0, 15}).
-			MatchCTMark(gatewayCTMark, nil).
-			Action().GotoTable(l2ForwardingCalcTable).
+			MatchDstIPNet(localSubnet).
+			Action().GotoTable(nextTable).
 			Cookie(c.cookieAllocator.Request(category).Raw()).
 			Done(),
-		// Forward the packet to L2ForwardingCalc table if it is sent to the Node IP(not to the host gateway). Since
-		// the packet is using the host gateway's MAC as dst MAC, it will be sent out from "antrea-gw0". This flow
-		// entry is to avoid SNAT on such packet, otherwise the source and destination IP are the same.
-		c.pipeline[l3ForwardingTable].BuildFlow(priorityLow).
+		// This flow is for the traffic to the local Node IP.
+		l3FwdTable.BuildFlow(priorityNormal).
 			MatchProtocol(binding.ProtocolIP).
 			MatchRegRange(int(marksReg), markTrafficFromLocal, binding.Range{0, 15}).
 			MatchDstIP(nodeIP).
-			Action().GotoTable(l2ForwardingCalcTable).
+			Action().GotoTable(nextTable).
 			Cookie(c.cookieAllocator.Request(category).Raw()).
 			Done(),
-		// Forward the packet to L2ForwardingCalc table if it is a packet sent to a local Pod. This flow entry has a
-		// low priority to avoid overlapping with those packets received from tunnel port.
-		c.pipeline[l3ForwardingTable].BuildFlow(priorityLow).
+		// This flow is for the return traffic of connections to a local
+		// Pod through the gateway interface (so gatewayCTMark is set).
+		// For example, the return traffic of a connection from an IP
+		// address (not the Node's management IP or gateway interface IP)
+		// of the local Node to a local Pod. It might also catch the
+		// Service return traffic from a local server Pod, but when the
+		// client of Service requests is a local Pod or the gateway
+		// interface, the return traffic is also caught by the flow
+		// matching the local subnet anyway.
+		l3FwdTable.BuildFlow(priorityNormal).
```

Comment:
I thought you planned to remove this flow, for the return packet of a Service from local Pod is always using gw0's IP as the destination on Windows, and it is already covered by the one line 1246.

Reply:
What you think about: 
```
                // For example, the return traffic of a connection from an IP
		// address (not the Node's management IP or gateway interface IP)
		// of the local Node to a local Pod
```

Reply:
In theory, such packet might exist, and it should require to enable IP-Forwarding on the interface where the source IP is configured. But according to our current configurations on Windows, the connection carrying gatewayCTMark always enters OVS from gw0, the source IP of the request (or the destination in the reply direction) is also gw0's IP address.

Reply:
I have no strong preferrence. The advantage to leave this flow is to make the SNAT flows more generic and extensible, while the disadvantage is the flow covered by another on in current Windows implementation.

Reply:
Yes, after the discussion with you, I also thought maybe IP forwarding will impact where the packets enter the pipeline. I still keep it because I plan to reuse the flows for Linux - for SNAT/egress policies. Probably let me add some comments to explain on Windows the flow might not be needed, as the traffic will not be forwarded to gw0 anyway, when IP forwarding is not enabled.